olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp

```diff
@@ -19,19 +19,107 @@
+#include "repositories/PartitionsRepository.h"
+
 namespace olp {
 namespace dataservice {
 namespace read {
 
+namespace {
+constexpr auto kLogTag = "VolatileLayerClientImpl";
+} // namespace
+
 VolatileLayerClientImpl::VolatileLayerClientImpl(
     olp::client::HRN catalog, std::string layer_id,
-    olp::client::OlpClientSettings client_settings) {}
+    olp::client::OlpClientSettings client_settings)
+    : catalog_(std::move(catalog)),
+      layer_id_(std::move(layer_id)),
+      client_settings_(std::move(client_settings)),
+      pending_requests_(std::make_shared<PendingRequests>()) {}
 
-VolatileLayerClientImpl::~VolatileLayerClientImpl() = default;
+VolatileLayerClientImpl::~VolatileLayerClientImpl() {
+  pending_requests_->CancelPendingRequests();
+}
 
 olp::client::CancellationToken VolatileLayerClientImpl::GetData(
     DataRequest data_request, Callback callback) {
-  return olp::client::CancellationToken{[]() {}};
+  auto catalog = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+
+  auto get_data = [=](DataRequest data_request,
+                      olp::client::CancellationContext context,
+                      Callback callback) {
+    if (!data_request.GetDataHandle()) {
+      auto response = repository::PartitionsRepository::GetPartitionById(
+          catalog, layer_id, context, data_request, client_settings);
+
+      if (!response.IsSuccessful()) {
+        if (callback) {
+          callback(response.GetError());
+        }
+        return;
+      }
+
+      const auto& partitions = response.GetResult().GetPartitions();
+      if (partitions.empty()) {
+        OLP_SDK_LOG_INFO_F(kLogTag, "Partition %s not found",
+                           data_request.GetPartitionId()
+                               ? data_request.GetPartitionId().get().c_str()
+                               : "<none>");
+        if (callback) {
+          callback(client::ApiError(client::ErrorCode::NotFound,
+                                    "Partition not found"));
+        }
+        return;
+      }
+
+      data_request.WithDataHandle(partitions.front().GetDataHandle());
+    }
+
+    auto blob_response = repository::DataRepository::GetBlobData(
+        catalog, layer_id, "volatile-blob", data_request, context,
+        client_settings);
+
+    if (callback) {
+      callback(blob_response);
+    }
+  };
+
+  auto pending_requests = pending_requests_;
+  auto add_task = [&](DataRequest& data_request, Callback callback) mutable {
+    olp::client::CancellationContext context;
+    auto cancellation_token = olp::client::CancellationToken(
+        [context]() mutable { context.CancelOperation(); });
+
+    int64_t request_key = pending_requests_->GenerateRequestPlaceholder();
+    pending_requests->Insert(cancellation_token, request_key);
+
+    repository::ExecuteOrSchedule(&client_settings, [=]() mutable {
+      get_data(data_request, context, std::move(callback));
```

Comment:
A little bit to many copies of the callback.
Why do we need it? Can we not just capture it?

Reply:
There is no easy way to move capture it in lambda.

Reply:
As you already copy capture the callback you actually only move the copy.
At least in most of the cases. So if you already copied it you can work with the copy directly.